src/segment-loader.js

```diff
@@ -267,47 +267,64 @@ export default class SegmentLoader extends videojs.EventTarget {
     let oldPlaylist = this.playlist_;
     let segmentInfo = this.pendingSegment_;
 
-    if (this.mediaIndex !== null) {
-      // We reloaded the same playlist so we are in a live scenario
-      // and we will likely need to adjust the mediaIndex
-      if (oldPlaylist &&
-          oldPlaylist.uri === newPlaylist.uri) {
-        let mediaSequenceDiff = newPlaylist.mediaSequence - oldPlaylist.mediaSequence;
-
-        this.mediaIndex -= mediaSequenceDiff;
-
-        if (segmentInfo) {
-          segmentInfo.mediaIndex -= mediaSequenceDiff;
-
-          // we need to update the referenced segment so that timing information is
-          // saved for the new playlist's segment, however, if the segment fell off the
-          // playlist, we can leave the old reference and just lose the timing info
-          if (segmentInfo.mediaIndex >= 0) {
-            segmentInfo.segment = newPlaylist.segments[segmentInfo.mediaIndex];
-          }
-        }
+    this.playlist_ = newPlaylist;
+    this.xhrOptions_ = options;
 
-        this.syncController_.saveExpiredSegmentInfo(oldPlaylist, newPlaylist);
-      } else {
-        // We must "resync" the fetcher when we switch renditions
-        this.resyncLoader();
-      }
-    } else if (!this.hasPlayed_()) {
+    // when we haven't started playing yet, the start of the playlist is always
+    // our zero-time so force a sync update each time we get a new playlist
+    if (!this.hasPlayed_()) {
       newPlaylist.syncInfo = {
         mediaSequence: newPlaylist.mediaSequence,
         time: 0
       };
       this.trigger('syncinfoupdate');
     }
 
-    this.playlist_ = newPlaylist;
-    this.xhrOptions_ = options;
-
     // if we were unpaused but waiting for a playlist, start
     // buffering now
     if (this.mimeType_ && this.state === 'INIT' && !this.paused()) {
       return this.init_();
     }
+
+    if (!oldPlaylist || oldPlaylist.uri !== newPlaylist.uri) {
+      if (this.mediaIndex !== null) {
+        // we must "resync" the segment loader when we switch renditions and
+        // the segment loader is already synced to the previous rendition
+        this.resyncLoader();
+      }
+
+      // the rest of this function depends on `oldPlaylist` being defined
+      return;
+    }
+
+    // we reloaded the same playlist so we are in a live scenario
+    // and we will likely need to adjust the mediaIndex
+    let mediaSequenceDiff = newPlaylist.mediaSequence - oldPlaylist.mediaSequence;
+
+    log('mediaSequenceDiff', mediaSequenceDiff);
+
+    // update the mediaIndex on the SegmentLoader
+    // this is important because we can abort a request and this value must be
+    // equal to the last appended mediaIndex
+    if (this.mediaIndex !== null) {
+      this.mediaIndex -= mediaSequenceDiff;
```

Comment:
Yeah, it's probably fine to do that. The behavior at least will be the same as current.

But it may be an interesting idea for the future. Since we're in a place where we knowingly will fall back to playback-watcher, and the user will experience a pause in playback, maybe we should at least add a comment or TODO to consider taking action.